lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1527,6 +1527,60 @@ static void diagRecursivePropertyAccess(const Expr *E, const DeclContext *DC) {
   const_cast<Expr *>(E)->walk(walker);
 }
 
+/// Whether or not this closure captures self weakly
+static bool closureHasWeakSelfCapture(const AbstractClosureExpr *ACE) {
+  if (auto closureExpr = dyn_cast<ClosureExpr>(ACE)) {
+    if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
+      return selfDecl->getType()->is<WeakStorageType>();
+    }
+  }
+
+  return false;
+}
+
+/// Whether or not implicit self DREs in this closure are incorrect,
+/// and need to be looked-up manually. This applies to weak self
+/// closures in Swift 5 mode, where the AST is incorrect.
+static bool
+closureHasIncorrectImplicitSelfDecls(const AbstractClosureExpr *ACE) {
+  if (ACE->getASTContext().LangOpts.isSwiftVersionAtLeast(6)) {
+    return false;
+  }
+
+  return closureHasWeakSelfCapture(ACE);
+}
+
+// Returns true if this is an implicit self expr
+static bool isImplicitSelf(const Expr *E) {
+  auto *DRE = dyn_cast<DeclRefExpr>(E);
+
+  if (!DRE || !DRE->isImplicit())
+    return false;
+
+  ASTContext &Ctx = DRE->getDecl()->getASTContext();
+  return DRE->getDecl()->getName().isSimpleName(Ctx.Id_self);
+}
+
+// When inside a closure with incorrect implicit self decls
+// (any closure with a weak self capture, in Swift 5 mode),
+// we have to manually lookup what self refers to at the location
+// of a given DRE instead of using the decl in the AST (which is incorrect).
+static ValueDecl *valueDeclForDRE(const DeclRefExpr *DRE,
+                                  const AbstractClosureExpr *ACE) {
+  ASTContext &Ctx = DRE->getDecl()->getASTContext();
+
+  if (closureHasIncorrectImplicitSelfDecls(ACE) && isImplicitSelf(DRE)) {
+    auto lookupResult = ASTScope::lookupSingleLocalDecl(
```

Comment:
This is the key change in this PR -- since the AST for implicit self decls is incorrect in Swift 5, we have to manually look up what it actually refers to. 

We use this when validating each individual usage of implicit self meets the requirements of SE-0365, and also in `VarDeclUsageChecker` to prevent false-positive "variable 'self' is unused" warnings.

Reply:
@xwu @rjmccall This is the key issue right here - my understanding was that the compiler should warn in cases like this that an incorrect self has been used instead of attempting to pretend that some other declaration has been picked (and in Swift 6 we should actually start picking correct `self` from local context). If the language group is okay with this approach I don't have any objections.

Reply:
@xwu or @rjmccall, mind taking a look? We could use some input on this.

Reply:
Thanks for the ping. This lgtm but it wouldn’t be my call on the implementation side.

Reply:
> That warning shouldn't be pretending that self was actually something different, but it does need to compare what self would be in Swift 6 with what we're using in Swift <6 to see if there's actually a problem.

The changes in this PR replace `self` declaration for diagnostic purposes, so no warning is going to be produced for non-escaping closures that picked an "incorrect" self, which is wrong - if I understand the outcome of the language group discussion correctly - we want to produce a tailored warning about behavior change between 5 and 6 in cases like that instead of "unused self" warning, to do so the implementation here would have to be reworked.

Reply:
> no warning is going to be produced for non-escaping closures that picked an "incorrect" self

With the implementation in this PR, the compiler does emit a warning here:

```swift
doVoidStuffNonEscaping { [weak self] in 
  // warning: call to method 'method' in closure requires explicit use of 'self' to make capture semantics explicit
  method()
}
```

and the following examples compile successfully with no warnings, since they comply with the behavior defined in SE-0365:

```swift
doVoidStuffNonEscaping { [weak self] in 
  // In Swift 5.7 we get a confusing "value 'self' was defined but never used; consider replacing with boolean test"
  // warning here. We would not emit this warning in Swift 5.8.
  guard let self else { return } 
  
  // Already compiles in Swift 5.7
  method()
}
```

```swift
doVoidStuff { [weak self] in 
  // In Swift 5.7 we get a confusing "value 'self' was defined but never used; consider replacing with boolean test"
  // warning here. We would not emit this warning in Swift 5.8.
  guard let self else { return }
  
  // Does not compile in Swift 5.7, now compiles without a warning
  method()
}
```

In both of these cases that would now compile successfully in Swift 5 mode, @xedin is correct the underlying AST is a bit unusual -- the implicit self DRE in the `method()` call does not actually refer to the `self` decl defined in the guard statement (which is why we have to do the additional lookup here in `MiscDiagnostics`). We can't fix this until Swift 6 mode, since fixing it is source-breaking.

The user-facing behavior of the compiler is still correct, though, even in Swift 5 mode. Implicit self calls are rejected unless they are preceded by a condition that unwraps self (e.g. `guard let self else { return }`).

So the fundamental question here seems to be whether it's acceptable for the AST to look one way but _behave_ as if it was something else (e.g. the AST that would be generated in Swift 6).

Reply:
>So the fundamental question here seems to be whether it's acceptable for the AST to look one way but behave as if it was something else (e.g. the AST that would be generated in Swift 6).

I don't think that's what the implementation is really doing. It's trying to approximate what `self` would refer to in Swift 6 in order to produce the `.warnUntilSwiftVersion(6)` diagnostic. I can see the argument for still producing a warning in the case where the behavior will "change" which doesn't currently produce a warning in this PR, but as Cal mentioned, from the user's perspective the behavior won't actually change in Swift 6.